File: algorithms/code/leetcode/lc234_palindrome_linked_list/lc234_palindrome_linked_list.py
# https://leetcode.com/problems/palindrome-linked-list/

from __future__ import annotations

import unittest


class ListNode:
    def __init__(
        self, val: int = 0, following: ListNode | None = None
    ) -> None:
        self.val = val
        self.next = following


class Solution:
    def isPalindrome(self, head: ListNode | None) -> bool:
        listed = []
        while head is not None:
            listed.append(head.val)
            head = head.next
        return all(
            listed[i] == listed[-1 - i] for i in range(len(listed) // 2)
        )


class TestCode(unittest.TestCase):
    def test_palindrome_odd(self) -> None:
        listed = ListNode(0)
        listed.next = ListNode(1)
        listed.next.next = ListNode(2)
        listed.next.next.next = ListNode(1)
        listed.next.next.next.next = ListNode(0)
        assert Solution().isPalindrome(listed)

    def test_palindrome_even(self) -> None:
        listed = ListNode(0)
        listed.next = ListNode(1)
        listed.next.next = ListNode(1)
        listed.next.next.next = ListNode(0)
        assert Solution().isPalindrome(listed)

    def test_not_palindrome_odd(self) -> None:
        listed = ListNode(0)
        listed.next = ListNode(1)
        listed.next.next = ListNode(2)
        listed.next.next.next = ListNode(8)
        listed.next.next.next.next = ListNode(0)
        assert not Solution().isPalindrome(listed)

    def test_not_palindrome_even(self) -> None:
        listed = ListNode(0)
        listed.next = ListNode(1)
        listed.next.next = ListNode(8)
        listed.next.next.next = ListNode(0)
        assert not Solution().isPalindrome(listed)

    def test_empty(self) -> None:
        assert Solution().isPalindrome(None)

    def test_single(self) -> None:
        assert Solution().isPalindrome(ListNode(0))

File: algorithms/code/leetcode/lc234_palindrome_linked_list/lc234_palindrome_linked_list.rs
// https://leetcode.com/problems/palindrome-linked-list/
// #easy

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

pub struct Solution;

impl Solution {
    pub fn is_palindrome(mut head: Option<Box<ListNode>>) -> bool {
        let mut listed = Vec::new();
        while let Some(node) = head {
            listed.push(node.val);
            head = node.next;
        }
        let count = listed.len();
        for i in 0..(count / 2) {
            if listed[i] != listed[count - 1 - i] {
                return false;
            }
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_palindrome_odd() {
        let n5: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 0, next: None }));
        let n4: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 1, next: n5 }));
        let n3: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 2, next: n4 }));
        let n2: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 1, next: n3 }));
        let n1: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 0, next: n2 }));
        let list = n1;
        assert!(Solution::is_palindrome(list));
    }

    #[test]
    fn test_palindrome_even() {
        let n4: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 0, next: None }));
        let n3: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 1, next: n4 }));
        let n2: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 1, next: n3 }));
        let n1: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 0, next: n2 }));
        let list = n1;
        assert!(Solution::is_palindrome(list));
    }

    #[test]
    fn test_not_palindrome_odd() {
        let n5: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 0, next: None }));
        let n4: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 8, next: n5 }));
        let n3: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 2, next: n4 }));
        let n2: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 1, next: n3 }));
        let n1: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 0, next: n2 }));
        let list = n1;
        assert!(!Solution::is_palindrome(list));
    }

    #[test]
    fn test_not_palindrome_even() {
        let n4: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 0, next: None }));
        let n3: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 8, next: n4 }));
        let n2: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 1, next: n3 }));
        let n1: Option<Box<ListNode>> = Some(Box::new(ListNode { val: 0, next: n2 }));
        let list = n1;
        assert!(!Solution::is_palindrome(list));
    }

    #[test]
    fn test_empty() {
        assert!(Solution::is_palindrome(None));
    }

    #[test]
    fn test_single() {
        assert!(Solution::is_palindrome(Some(Box::new(ListNode {
            val: 0,
            next: None,
        }))));
    }
}
